### DISCUSSION TITLE
[GraphSage]: Is the size always the same between number of layers in SAGE and len(size) in NeighborSampler?

### DISCUSSION THREAD

**Discussion by udothemath** [pyg-team/pytorch_geometric#3799] [Q&A]
There are three SAGEConv examples provided by pytorch geometric. I am wondering is the size of number of neighbors in dataloader always the same as number of SAGEConv layer?

For [example 1: reddit](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/reddit.py), num_neighbors=[25, 10] in Neighborloader, and there are two SAGEConv in SAGE(torch.nn.Module) class.
![image](https://user-images.githubusercontent.com/10674490/147989499-41073084-3e0e-4d78-9bcc-65193810cff4.png)
![image](https://user-images.githubusercontent.com/10674490/147989523-0e3dfe2f-f459-4861-8b34-7299e8174a54.png)

Similar setting for example 2 and 3.

I can understand the SAGEConv concept of sampling from the node's neighbor, but I don't understand how this idea is implemented in neural network. 

In graph, each circle means a node.
![source: https://www.arxiv-vanity.com/papers/2103.16329/](https://user-images.githubusercontent.com/10674490/147990287-47b1c6a6-7b24-4066-b03f-8e210eb38222.png)

In neural network, each circle means a neuron.
![source: https://commons.wikimedia.org/wiki/File:Neural_network.svg](https://user-images.githubusercontent.com/10674490/147990339-b48111ae-5ac0-4f7f-a4c6-b6b1179b0fae.png)

How to put these two together? Is there any good reference/visualization to explain the concept? Thanks for your help.

--
For [example 2: ogbn_products_sage.](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_products_sage.py), sizes=[15, 10, 5] in NeighborSampler, and there are three SAGEConv in SAGE model (num_layers=3).

For [example 3: graph_sage_unsup](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/graph_sage_unsup.py), sizes=[10, 10] in NeighborSampler, and num_layers=2 in SAGE model.

**Comment by rusty1s** [comment]
You are right. The number of hops (`len(num_neighbors)`) and the number of layers should match. In your figure, the squared gray boxes denote the application of a GNN layer, and for each round of sampling, we apply a different GNN layer.

A node in the computation graph actually refers to a set of neuron since each node is described by a set of features. For a GNN with input, hidden and output channels equals to 1, a circle describes both a node and a single neuron. Hope this clarifies your issues.

**Reply by imethanlee** [reply]
Should `num_neighbors` in `train_loader` and `test_loader` be the same (I mean the length and the entries of `num_neighbors`)? I checked the code in [https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_products_sage.py#L23](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_products_sage.py#L23) and [https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_products_gat.py#L24](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/ogbn_products_gat.py#L24) and found that they are not the same. Is there any intuition on how to set the `num_neighbors` in `test_loader` based on `train_loader`?

**Reply by rusty1s** [reply]
The number of neighbors does not need to match. You can think of this argument as some form of dropout, and increasing the number of neighbors to sample during inference reduces variance in estimation, and usually leads to better model predictions.

**Reply by imethanlee** [reply]
Thanks for the prompt reply on the entries for the `num_neighbors` in `NeighborLoader` (which is actually the argument `size` in `NeighborSampler).` Just one more question to go for the length of `num_neighbors`. You can see from the codes I mentioned in my last post, the `num_neighbors` in `train_loader` is [15, 10, 5], but in `test_loader` is [-1]. Does it mean that in training, we actually consider 1-, 2-, and 3-hop neighbors, but in testing, we only consider 1-hop neighbors? If I want to consider 1-, 2-, and 3-hop in testing, should I change `num_neighbors` in `test_loader` to [-1, -1, -1]?

**Reply by rusty1s** [reply]
Ah sorry, I just have checked better :) During inference, we actually utilize layer-wise computation of node embeddings (see the `inference` implementation). That is, we sample the 1-hop neighborhood for each node, update the hidden node representations for all of them, and then proceed with the next layer. This is actually far more efficient than sampling `[-1, -1, -1]` neighbors since we avoid the exponential increase in neighborhood sizes. Let me know if this clarifies your doubts!

**Reply by imethanlee** [reply]
Ah, this clarifies all my doubts. Thanks a lot :)

**Reply by imethanlee** [reply]
Well actually there is one more question. I am using `NeighborLoader` instead of `NeighborSampler` now, how can I access the node id within each batch?

Perhaps I can make my question more specific. If I use `NeighborLoader` to form my `subgraph_loader`, how can I do efficient inference (by sampling only 1-hop neighbor and update all node hidden representation layer-by-layer as you said in your last reply)? In `NeighborSampler` we have `n_id` to access directly but in `NeighborLoader` I fail to figure out where to access `n_id.`

**Reply by rusty1s** [reply]
You can add it as an attribute to `data`, e.g.: `data.n_id = torch.arange(data.num_nodes)`. That makes it available in the mini-batches of `NeighborLoader` as well.

**Reply by jylink** [reply]
I know that the length of num_neighbors and the number of layers in SAGEConv should match, but I can't relate them to the algorithm in the GraphSAGE paper.

For the [reddit](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/reddit.py) example, each batch contains 1024 sampled nodes (noted as b3) and their 1-hop (b2) and 2-hop neighbors (b1), and then b3&b2&b1 are passed to the two SAGEConv layers, is that right?

But according to the paper, shouldn't we pass b3&b2&b1 to the first SAGEConv layer and b3&b2 to the second SAGEConv layer?

Also, what is the length of NeighborLoader? How does it determine the number of batches per iteration?

**Reply by rusty1s** [reply]
Good question. You are right that one usually does not need to pass in the full set of nodes to each layer. We do this for simplicity: We do not need to manage different adjacency matrices across layers, do not need to handle bipartite message passing, and only need to slice the node feature matrix  once at the end of the iteration. This simplicity comes with a small increase in runtime since we are actually computing unused node embeddings.

The length of the loader is given by len(seed_nodes) / batch_size.
